### ISSUE TITLE
Tighter parameter decorator checking with `--experimentalDecorators`

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/TypeScript#52435]
```ts
// @experimentalDecorators: true
export declare const Inject:
  (entity: Function, dataSource?: string) =>
    (target: object, key: string | symbol, index?: number) => void;

export class Foo {}

export class C {
  constructor(@Inject(Foo) x: any) {  }
}
```

Previously: no error
In `main`/5.0:

```
Unable to resolve signature of parameter decorator when called as an expression.
  Argument of type 'undefined' is not assignable to parameter of type 'string | symbol'
```

**Comment by DanielRosenwasser** [comment]
Was this...possibly always a bug with `--experimentalDecorators`?

**Comment by DanielRosenwasser** [comment]
Yes, it seems this was a bug in type-checking parameter decorators under `--experimentalDecorators` with TypeScript 4.9 and prior.

In #50820, we fixed up the way we type-checked decorators. This catches some issues, like decorators which expect a defined `key`; however, parameter decorators don't pass a key. As a result, libraries with dependency injection decorator functions like inversify, NestJS, and others (often named `inject`, `Inject`, and `InjectRepository`) might need to be corrected to accept an `undefined` key (or to have an overload which accepts an `undefined` key).

**Comment by rbuckton** [comment]
`inversify`'s `@inject` works fine starting in 6.0.1. `@nestjs/common`'s `@inject` needs to be updated, however.

**Comment by kevinlul** [comment]
Is tsyringe affected?

**Comment by DanielRosenwasser** [comment]
[I think so.](https://github.com/microsoft/tsyringe/blob/f738999f3058b223bdb81dada0164a358db2460e/src/decorators/inject.ts#L11)

**Comment by DanielRosenwasser** [comment]
I filed an issue at NestJS over at https://github.com/microsoft/TypeScript/issues/52435. I'm not sure how active TSyringe is.

**Comment by Evaldoes** [comment]
I was able to solve this problem by adding these parameters to the `tsconfig.json` file:

I used `inversify: 6.0.1` version

**NOTE:** when I inserted it at the end of the file it didn't work, only when I put it at the beginning

```
{
  "compilerOptions": {
      "emitDecoratorMetadata": true,
      "experimentalDecorators": true,
      "allowJs": true,
  }
}
```

I hope I can help :smile:
